File: app/src/main/java/com/example/pearvideoclient/follow/addfollow/AddFollowCategoryAdapter.java
package com.example.pearvideoclient.follow.addfollow;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.example.pearvideoclient.R;
import com.example.pearvideoclient.databinding.AdapterDomainItemBinding;
import com.example.pearvideoclient.entity.DomainListBean;

import java.util.List;
import java.util.Objects;

/**
 * @Description: java类作用描述
 * @Author: jiazhu
 * @Date: 2019-09-02 17:32
 * @ClassName: AddFollowCategoryAdapter
 */
public class AddFollowCategoryAdapter extends RecyclerView.Adapter<AddFollowCategoryAdapter.DomainViewHolder> {
    private List<DomainListBean.DomainBean> mDomainLists;
    @Nullable
    private final DomainClickCallback mDomainClickCallback;

    public AddFollowCategoryAdapter(@Nullable DomainClickCallback domainClickCallback) {
        mDomainClickCallback = domainClickCallback;
    }

    public List<DomainListBean.DomainBean> getDomainLists() {
        return this.mDomainLists;
    }

    public void setDomainLists(final List<DomainListBean.DomainBean> domainLists) {
        if (this.mDomainLists == null) {
            this.mDomainLists = domainLists;
            notifyItemRangeChanged(0, domainLists.size());
        } else {
            DiffUtil.DiffResult result = DiffUtil.calculateDiff(new DiffUtil.Callback() {
                @Override
                public int getOldListSize() {
                    return AddFollowCategoryAdapter.this.mDomainLists.size();
                }

                @Override
                public int getNewListSize() {
                    return domainLists.size();
                }

                @Override
                public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                    return AddFollowCategoryAdapter.this.mDomainLists.get(oldItemPosition).getDomainId().equals(domainLists.get(newItemPosition).getDomainId());
                }

                @Override
                public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                    DomainListBean.DomainBean newDomain = domainLists.get(newItemPosition);
                    DomainListBean.DomainBean oldDomain = domainLists.get(oldItemPosition);
                    return Objects.equals(newDomain.getDomainId(), oldDomain.getDomainId()) &&
                            Objects.equals(newDomain.getName(), oldDomain.getName());
                }
            });
            this.mDomainLists = domainLists;
            result.dispatchUpdatesTo(this);
        }
    }

    @NonNull
    @Override
    public DomainViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        AdapterDomainItemBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()),
                R.layout.adapter_domain_item, parent, false);
        binding.setCallback(mDomainClickCallback);
        return new DomainViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull DomainViewHolder holder, int position) {
        holder.mBinding.setDomain(mDomainLists.get(position));
    }

    @Override
    public int getItemCount() {
        return mDomainLists == null ? 0 : mDomainLists.size();
    }

    static class DomainViewHolder extends RecyclerView.ViewHolder {
        final AdapterDomainItemBinding mBinding;

        public DomainViewHolder(AdapterDomainItemBinding binding) {
            super(binding.getRoot());
            this.mBinding = binding;
        }
    }
}

File: app/src/main/java/com/example/pearvideoclient/follow/addfollow/AddFollowCategoryAdapter.kt
package com.example.pearvideoclient.follow.addfollow

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.example.pearvideoclient.R
import com.example.pearvideoclient.databinding.AdapterDomainItemBinding
import com.example.pearvideoclient.entity.DomainListBean

/**
 * @Description: java类作用描述
 * @Author: jiazhu
 * @Date: 2019-09-02 17:32
 * @ClassName: AddFollowCategoryAdapter
 */
class AddFollowCategoryAdapter(private val mDomainClickCallback: DomainClickCallback?)
    : RecyclerView.Adapter<AddFollowCategoryAdapter.DomainViewHolder>() {
    var domainLists: List<DomainListBean.DomainBean>? = null
        set(domainLists) = if (this.domainLists == null) {
            field = domainLists
            notifyItemRangeChanged(0, domainLists!!.size)
        } else {
            val result = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize(): Int {
                    return this@AddFollowCategoryAdapter.domainLists!!.size
                }

                override fun getNewListSize(): Int {
                    return domainLists!!.size
                }

                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    return this@AddFollowCategoryAdapter.domainLists!![oldItemPosition].domainId == domainLists!![newItemPosition].domainId
                }

                override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    val newDomain = domainLists!![newItemPosition]
                    val oldDomain = domainLists[oldItemPosition]
                    return newDomain.domainId == oldDomain.domainId && newDomain.name == oldDomain.name
                }
            })
            field = domainLists
            result.dispatchUpdatesTo(this)
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DomainViewHolder {
        val binding = DataBindingUtil.inflate<AdapterDomainItemBinding>(LayoutInflater.from(parent.context),
                R.layout.adapter_domain_item, parent, false)
        binding.callback = mDomainClickCallback
        return DomainViewHolder(binding)
    }

    override fun onBindViewHolder(holder: DomainViewHolder, position: Int) {
        holder.mBinding.domain = domainLists!![position]
    }

    override fun getItemCount(): Int {
        return if (domainLists == null) 0 else domainLists!!.size
    }

    class DomainViewHolder(val mBinding: AdapterDomainItemBinding) : RecyclerView.ViewHolder(mBinding.root)
}
